File: wrappers/javascript/anoncreds-shared/src/api/RevocationStatusList.ts
import type { ObjectHandle } from '../ObjectHandle'
import type { JsonObject } from '../types'

import { AnoncredsObject } from '../AnoncredsObject'
import { anoncreds } from '../register'

import { RevocationRegistryDefinition } from './RevocationRegistryDefinition'
import { pushToArray } from './utils'

export type CreateRevocationStatusListOptions = {
  revocationRegistryDefinitionId: string
  revocationRegistryDefinition: RevocationRegistryDefinition | JsonObject
  issuerId: string
  timestamp?: number
  issuanceByDefault: boolean
}

export type UpdateRevocationStatusListTimestampOptions = {
  timestamp: number
}

export type UpdateRevocationStatusListOptions = {
  revocationRegistryDefinition: RevocationRegistryDefinition | JsonObject
  timestamp?: number
  issued?: Array<number>
  revoked?: Array<number>
}

export class RevocationStatusList extends AnoncredsObject {
  public static create(options: CreateRevocationStatusListOptions) {
    let revocationStatusListHandle
    const objectHandles: ObjectHandle[] = []
    try {
      const revocationRegistryDefinition =
        options.revocationRegistryDefinition instanceof RevocationRegistryDefinition
          ? options.revocationRegistryDefinition.handle
          : pushToArray(
              RevocationRegistryDefinition.fromJson(options.revocationRegistryDefinition).handle,
              objectHandles
            )

      revocationStatusListHandle = anoncreds.createRevocationStatusList({
        ...options,
        revocationRegistryDefinition,
      }).handle
    } finally {
      objectHandles.forEach((handle) => handle.clear())
    }
    return new RevocationStatusList(revocationStatusListHandle)
  }

  public static fromJson(json: JsonObject) {
    return new RevocationStatusList(anoncreds.revocationStatusListFromJson({ json: JSON.stringify(json) }).handle)
  }

  public updateTimestamp(options: UpdateRevocationStatusListTimestampOptions) {
    const updatedRevocationStatusList = anoncreds.updateRevocationStatusListTimestampOnly({
      timestamp: options.timestamp,
      currentRevocationStatusList: this.handle,
    })

    this.handle = updatedRevocationStatusList
  }

  public update(options: UpdateRevocationStatusListOptions) {
    const objectHandles: ObjectHandle[] = []
    try {
      const revocationRegistryDefinition =
        options.revocationRegistryDefinition instanceof RevocationRegistryDefinition
          ? options.revocationRegistryDefinition.handle
          : pushToArray(
              RevocationRegistryDefinition.fromJson(options.revocationRegistryDefinition).handle,
              objectHandles
            )

      const updatedRevocationStatusList = anoncreds.updateRevocationStatusList({
        ...options,
        revocationRegistryDefinition,
        currentRevocationStatusList: this.handle,
      })

      this.handle = updatedRevocationStatusList
    } finally {
      objectHandles.forEach((handle) => handle.clear())
    }
  }
}

File: wrappers/kotlin/src/commonMain/kotlin/api/RevocationStatusList.kt
package anoncreds.api

import anoncreds.Anoncreds
import anoncreds_rs.*
import kotlinx.cinterop.*
import platform.posix.int32_tVar

class RevocationStatusList
private constructor(
    handle: ObjectHandle
) : AnoncredsObject(handle) {

    companion object{

        fun create(
            revocationRegistryDefinitionId: String,
            revocationRegistryDefinition: RevocationRegistryDefinition,
            issuerId: String,
            issuanceByDefault: Boolean,
            timestamp: Long,
        ): RevocationStatusList {
            memScoped {
                val listPointer = alloc<ObjectHandleVar>()

                val errorCode = anoncreds_create_revocation_status_list(
                    revocationRegistryDefinitionId,
                    revocationRegistryDefinition.handle,
                    issuerId,
                    timestamp,
                    if(issuanceByDefault) 1 else 0,
                    listPointer.ptr
                )

                Anoncreds.assertNoError(errorCode)

                val listHandle = listPointer.value

                return RevocationStatusList(listHandle)
            }
        }

        fun fromJson(json: String): RevocationStatusList {
            return fromJson(
                json,
                ::RevocationStatusList,
                ::anoncreds_revocation_status_list_from_json
            )
        }
    }

    fun updateTimestamp(timestamp: Long){
        memScoped {
            val listPointer = alloc<ObjectHandleVar>()
            val errorCode = anoncreds_update_revocation_status_list_timestamp_only(
                timestamp,
                handle,
                listPointer.ptr
            )

            Anoncreds.assertNoError(errorCode)

            handle = listPointer.value
        }
    }

    fun update(
        revocationRegistryDefinition: RevocationRegistryDefinition,
        timestamp: Long,
        issued: List<Int>,
        revoked: List<Int>
    ){
        memScoped {
            val listPointer = alloc<ObjectHandleVar>()

            val issuedArray = allocArray<int32_tVar>(issued.size)
            for(i in 0 .. issued.size){
                issuedArray[i] = issued[i]
            }
            val issuedC = cValue<FfiList_i32>{
                this.count = issued.size.convert()
                this.data = issuedArray
            }

            val revokedArray = allocArray<int32_tVar>(revoked.size)
            for(i in 0 .. revoked.size){
                revokedArray[i] = revoked[i]
            }
            val revokedC = cValue<FfiList_i32>{
                this.count = revoked.size.convert()
                this.data = revokedArray
            }

            val errorCode = anoncreds_update_revocation_status_list(
                timestamp,
                issuedC,
                revokedC,
                revocationRegistryDefinition.handle,
                handle,
                listPointer.ptr
            )

            Anoncreds.assertNoError(errorCode)

            handle = listPointer.value
        }
    }
}